python/ray/rllib/test/test_checkpoint_restore.py

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env python
+
+import ray
+
+from ray.rllib.dqn import DQN, DEFAULT_CONFIG
+cls = DQN
+
+ray.init()
+config = DEFAULT_CONFIG.copy()
+config["num_sgd_iter"] = 5
+config["episodes_per_batch"] = 100
+config["timesteps_per_batch"] = 1000
+alg1 = cls('CartPole-v0', config)
+
+print("Test action", alg1.compute_action([0, 0, 0, 0]))
+
+res = alg1.train()
+while res.episode_reward_mean < 180:
```

Comment:
Let's do the following: for each of the algorithms, run the training/checkpoint loop for a little bit and then fix a set of observations, try to evaluate the last feature layer of the policy (with floating point values) both on the algorithm instance and also on the loaded checkpoint and make that they are the same.

It's fine if they are not at the moment and stuff needs to be fixed (i.e. we can have an expected failure), but it will be good to know if it works.

Reply:
Instead of testing that the network layers are the same, maybe we should test that the predictions on some inputs are the same (to catch other missing things like the observation filter).

Reply:
Good suggestion. Updated. There are a few issues:

- A3C / PolicyGradient have stochastic policies (we can fix this)
- ES will crash during `get_trainable_flat` due to `ValueError: Fetch argument <tf.Tensor 'concat:0' shape=(67586,) dtype=float32> cannot be interpreted as a Tensor. (Tensor Tensor("concat:0", shape=(67586,), dtype=float32) is not an element of this graph.)`. This is probably a bug in tf_util (sigh) that doesn't respect the current graph.

These are probably not blockers for the tutorial though.

Reply:
Using this PR I think I was able to instantiate multiple ES "algorithm" objects and train each of them.

How do you reproduce the crash?

Also, I don't think the tf_utils file was designed to support having multiple networks. The solution is probably to stop using those tf_utils.